Proposal for tenant retention (#3638)
cortexproject/cortex: docs/proposals/tenant_retention.md

- Date: January 2021
- Status: Proposed

## Retention of tenant data

## Problem

Metric data is growing over time per-tenant, at the same time, the value of data decreases. We want to have a retention policy like prometheus does. In Cortex, data retention is typically achieved via a bucket policy. However, this has two main issues:

1. Not every backend storage support bucket policies
2. Bucket policies don't easily allow a per-tenant custom retention

## Background

### tenants
When using blocks storage, Cortex stores tenant’s data in object store for long-term storage of blocks, tenant id as part of the object store path. We discover all tenants via scan the root dir of bucket.

### runtime config
Using the "overrides" mechanism (part of runtime config) already allows for per-tenant settings. See [runtime-configuration-file](https://cortexmetrics.io/docs/configuration/arguments/#runtime-configuration-file) for more details. Using it for tenant retention would fit nicely. Admin could set per-tenant retention here, and also have a single global value for tenants that don't have custom value set.

## Proposal

### retention period field

> We propose to introduce just one new field `RetentionPeriodPerTenant` in the Limits struct(defined at pkg/util/validation/limits.go).

Reviewer: All fields in the `Limits` struct are per-tenant so no need to make it explicit.

```suggestion
We propose to introduce just one new field `RetentionPeriod` in the Limits struct(defined at pkg/util/validation/limits.go).
```

> `RetentionPeriodPerTenant` setting how long historical metric data retention period per-tenant. `0` is disable.

Reviewer 2: I assume `RetentionPeriod` and `RetentionPeriodPerTenant` are the same. We should use just one name here.

Author: Yes, they are same. Both should be `RetentionPeriod`.

> As a runtime config, the settings is reload period (which defaults to 10 seconds), so we can update the settings onfly.

Reviewer: ```suggestion
Runtime config is reloaded periodically (defaults to 10 seconds), so we can update the retention settings on-the-fly.
```

For each tenant, if a tenant-specific *runtime_config* value exists, it will be used directly, otherwise, if a default *limits_config* value exists, then the default value will be used; If neither exists, do nothing.

### Implementation

> A BlocksCleaner within the Compactor run periodically(which defaults to 1 hour) and the retention logic will insert into it. The logic should compare retention value to block `maxTime` and blocks that match `maxTime < now - retention` will be marked for delete.

Reviewer 2: ```suggestion
A BlocksCleaner within the Compactor run periodically (which defaults to 15 minutes) and the retention logic will insert into it. The logic should compare retention value to block `maxTime` and blocks that match `maxTime < now - retention` will be marked for delete.
```

---

- Author: [Allenzhli](https://github.com/Allenzhli)
- Date: January 2021
- Status: Proposed

## Retention of tenant data

## Problem

Metric data is growing over time per-tenant, at the same time, the value of data decreases. We want to have a retention policy like prometheus does. In Cortex, data retention is typically achieved via a bucket policy. However, this has two main issues:

1. Not every backend storage support bucket policies
2. Bucket policies don't easily allow a per-tenant custom retention

## Background

### tenant
When using blocks storage, Cortex stores tenant’s data in object store for long-term storage of blocks, tenant id as part of the object store path. We discover all tenants via scan the root dir of bucket.

### runtime config
Using the "overrides" mechanism (part of runtime config) already allows for per-tenant settings. See https://cortexmetrics.io/docs/configuration/arguments/#runtime-configuration-file for more details. Using it for tenant retention would fit nicely. Admin could set per-tenant retention here, and also have a single global value for tenants that don't have custom value set.

## Proposal

### retention component

> We propose to introduce an new runtime config component `retention`.

Reviewer: I don't think this is necessary. We can just have 1 new field in the `Limits` struct (defined at `pkg/util/validation/limits.go`) which allows to specify the per-tenant retention (0 to disable it).

Author: You are right.  I didn't understand the runtime config correctly.

We can store per-tenant settings and global setting in the retention component. As a runtime config, the settings is reload period (which defaults to 10 seconds), so we can update the settings onfly.

Per-tenant retention configuration will be applyed for in follow order:

* tenant specified value
* global value

For each tenant, if a tenant-specific value exists, it will be used directly, otherwise, if a global value exists, then the global value will be used; If neither exists, do nothing.

### Implementation

> The runtime config will add a retention component which include a dictionary to store per-tenant retention value and global retention value.

Reviewer: See above. No need for a new dictionary/map. We can just reuse `Limits`, which is the way we currently implement any per-tenant config.

Author: Same as above. I will reuse the `Limits`.

> Compactor will check per-tanant retention value and compare to meta info of blocks. The blocks that match `maxTime < now - retention` will be marked for delete.

Reviewer: FYI, we have a blocks cleaner component within the compactor. That could be the right place to add the retention logic.

Author: I agree with you. I tend to insert the logic in function `cleanUser`.

> ## Compactor

Reviewer: I'm not sure what's the purpose of this section. Is it to explain how things are working today in the compactor?

Author: I want to explain how to implement blocks delete.

---
title: "Retention of Tenant Data from Blocks Storage"
linkTitle: "Retention of Tenant Data from Blocks Storage"
weight: 1
slug: tenant-retention
---

- Author: [Allenzhli](https://github.com/Allenzhli)
- Date: December 2020
- Status: Proposed

## Retention of tenant data

## Problem

> In Cortex, we are missing retention of blocks storage. The metric data is growing over time per-tenant, at the same time, the value of data decreases. We want to support retention time of metric data like prometheus does.

Reviewer: This is not entire correct. Data retention is typically achieved be achieved via a bucket policy. However, we should explain this has two main issues:

1. Not every backend storage support bucket policies
2. Bucket policies don't easily allow a per-tenant custom retention

Author: I agree with you. I will update this.

## Background

When using blocks storage, Cortex stores tenant’s data in object store for long-term storage of blocks, tenant id as part of the object store path. We discover all tenants via scan the root dir of bucket.

## Proposal

### API Endpoints

#### POST /compactor/retention_tenant

> We propose to introduce an `/compactor/retention_tenant` API endpoint to set data retention for the tenant.

Reviewer: I wouldn't introduce any new API. We should use the already existing mechanism used to set per-tenant overrides (via the runtime config "overrides").

Author: I will delete the api design.

While this endpoint works as an “admin” endpoint and should not be exposed directly to tenants, it still needs to know under which tenant it should operate.
For consistency with other endpoints this endpoint would therefore require an `X-Scope-OrgID` header and use the tenant from it.

It is safe to call “retention_tenant” multiple times.

#### GET /compactor/retention_tenant

To query the state of data retention, an endpoint will be available: `/compactor/retention_tenant`.
This will return the retention of current tenant.
For consistency with other endpoints this endpoint would therefore require an `X-Scope-OrgID` header and use the tenant from it.

### Implementation (asynchronous)

Compactor will implement both API endpoints.
Upon receiving the call to `/compactor/retention_tenant` endpoint, Compactor will initiate the retention by writing “retetion marker” objects for specific tenant to blocks bucket.

> Retention marker for the tenant will be an object stored under tenant prefix, eg. “/<tenant>/retention”.

Reviewer: I wouldn't use any retention marker. The per-tenant retention should be part of the config.

Author: I agree with you, this will be delete.
